tracy-client/src/lib.rs

```diff
@@ -1,23 +1,137 @@
+//! expose the data it collects in the background to that same network. Traces collected by Tracy
+//! may include source and assembly code as well.
 //!
-//! As thus, you may want make sure to only enable the `tracy-client` crate conditionally, via the
-//! `enable` feature flag provided by this crate.
+//! As thus, you may want make sure to only enable the `tracy-client` crate conditionally, via
+//! the `enable` feature flag provided by this crate.
 //!
-//! [Tracy profiler]: https://github.com/wolfpld/tracy
-#![cfg_attr(not(feature="enable"), allow(unused_imports, unused_variables))]
+//! In order to start tracing it is important that you first call the [`enable`] function first to
+//! initialize the client. The [`disable`] must not be called until it is guaranteed that there
+//! will be no more calls to any other APIs. This can be especially difficult to ensure if you have
+//! detached threads. It is valid to never call `disable` at all.
+//!
+//! # Features
+//!
+//! Refer to the [`sys`] crate for documentation on crate features. This crate re-exports all the
+//! features from [`sys`].
+#![cfg_attr(not(feature = "enable"), allow(unused_imports, unused_variables))]
 
 use std::alloc;
 use std::ffi::CString;
-use tracy_client_sys as sys;
+use std::sync::atomic::{AtomicUsize, Ordering};
+pub use sys;
+
+/// When enabling `Tracy` when it is already enabled, or when disabling when it is already disabled
+/// can cause applications to crash. I personally think it would be better if this was a sort-of
+/// reference counted kind-of thing so you could enable as many times as you wish and disable just
+/// as many times without any reprecursions. At the very least this could significantly help tests.
+///
+/// In order to do this we want to track 4 states that construct a following finite state machine
+///
+/// ```text
+///     0 = disabled  -----> 1 = enabling
+///         ^                    v
+///     3 = disabling <----- 2 = enabled
+/// ```
+///
+/// And also include a reference count somewhere in there. Something we can throw in a static would
+/// be ideal.
+///
+/// Sadly, I am not aware of any libraries which would make this easier, so rolling out our own
+/// things it is then!
+///
+/// Oh and it'll definitely involve spinning, getting some vertigo medication is advised.
+///
+/// ---
+///
+/// We will use a single Atomic to store this information. The 2 top-most bits will represent the
+/// state bits and the rest will act as a counter.
+static ENABLE_STATE: AtomicUsize = AtomicUsize::new(0);
+const REFCOUNT_MASK: usize = usize::max_value() >> 2;
+const STATE_STEP: usize = REFCOUNT_MASK + 1; // Move forward by 1 step in the FSM
+const STATE_DISABLED: usize = 0;
+const STATE_ENABLING: usize = STATE_DISABLED + STATE_STEP;
+const STATE_ENABLED: usize = STATE_ENABLING + STATE_STEP;
+const STATE_DISABLING: usize = STATE_ENABLED + STATE_STEP;
+const STATE_MASK: usize = STATE_DISABLING;
+
+pub fn enable() {
```

Comment:
TODO: these need documentation.

I'm also torn between making the enable and disable two distinct operations vs getting some sort of guard setup going here...

Reply:
A guard would make it much harder to fail to disable gracefully on exit, and is rustier.

Reply:
Right, OTOH, not calling `disable` at all is going to be more correct out of the box most of the time.

As soon as the process ends up with a detached thread running instrumented code, calling `disable` would lead to a crash. And Rust code generally _loves_ having detached threads running things...

Reply:
> And Rust code generally loves having detached threads running things...

This is certainly not true of any Rust I have ever had a hand in. Still, if failing to call `disable` is harmless, and calling it prematurely is unsound, that's a pretty good argument to leave it off the default path.